[mlir][Transforms] `GreedyPatternRewriteDriver`: Check for out-of-scope IR modifications

matthias-springer opened the pull request:
This commit adds an additional "expensive check" (only enabled with `MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS`) that looks for out-of-scope IR modifications.

`GreedyRewriteConfig::scope` specifies the `Region *` within which the greedy pattern rewrite operates. Operations that are out-of-scope are not added to the worklist. (This is already the case.) The new expensive check triggers a fatal error if:
* Op is inserted into out-of-scope region.
* Op is removed from out-of-scope region.
* Op is modified in out-of-scope region. (`scope->getParentOp()` may also be modified if the `applyPatternsAndFoldGreedily(Operation *, ...)` entry point is used.)

Note: The scope can be set in `GreedyRewriteConfig`. If no scope was specified by the user (which is the typical case), it is inferred from the parameters of the greedy pattern rewriter entry point.

This change also tightens the greedy pattern rewriter entry points and makes sure that the specified `scope` is an `IsolatedFromAbove` region. (We had such a check only for one of the entry points.)

joker-eph commented:
> Note: TileAllocation (ArmSME dialect) must now be a module pass because it modifies func.func ops (adds attributes). This is forbidden for function passes

I would think we'd allow this though. @Mogball WDYT?

Mogball commented:
> > Note: TileAllocation (ArmSME dialect) must now be a module pass because it modifies func.func ops (adds attributes). This is forbidden for function passes
> 
> I would think we'd allow this though. @Mogball WDYT?

I believe the restriction on nested passes is that the parent op is ready only, whilst sister ops cannot be read or written. Modifying the op itself is fine, so this pass can remain a function pass.

joker-eph commented:
I thought the same, but this isn't what we document on https://mlir.llvm.org/docs/PassManagement/#operation-pass

> Must not modify any state referenced or relied upon outside the current operation being operated on. This includes adding or removing operations from the parent block, **changing the attributes(depending on the contract of the current operation)/operands/results/successors of the current operation**.

matthias-springer commented:
I updated the pass management documentation in #77614 based on your comments. This PR now depends on #77614 and modifying the parent (owner) op of the scope region is now allowed. In practical terms: Greedy rewrites that are invoked on an operation `op` are now allowed to modify `op` (only modify, not remove).

Mogball commented:
> (only modify, not remove).

Can you clarify what this mean? Doesn't every pattern remove the op.

matthias-springer commented:
> Can you clarify what this mean? Doesn't every pattern remove the op.

Something like:
```c++
rewriter.updateRootInPlace(op, [&](){ op->setAttr(...); });
```

matthias-springer committed:
[mlir][Transforms] `GreedyPatternRewriteDriver`: Check for out-of-scope IR modifications

This commit adds an additional "expensive check" (only enabled with `MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS`) that looks for out-of-scope IR modifications.

`GreedyRewriteConfig::scope` specifies the `Region *` within which the greedy pattern rewrite operates. Operations that are out-of-scope are not added to the worklist. The new expensive check triggers an fatal error if:
* Op is inserted into out-of-scope region.
* Op is removed from out-of-scope region.
* Op is modified in out-of-scope region.

This change also tightens the greedy pattern rewriter entry points and makes sure that the specified `scope` is an `IsolatedFromAbove` region.

Note: `TileAllocation` (`ArmSME` dialect) must now be a module pass because it modifies `func.func` ops (adds attributes). This is forbidden for function passes (in which the scope of the greedy rewrite is set to the region of the function by default) because only function bodies are allowed to be modified. (TODO: Should we allow this? Is there something special about functions?)

matthias-springer commented:
Rebase, simplify implementation a bit and add more documentation.
